### PR TITLE
Report cloud

### PR DESCRIPTION
**Pull request by MallanagoudaB**
# Description
These recommendations are intended to promote code quality and team communication during software development. They cover a variety of topics, including ensuring that pull requests are submitted to the correct branch, documenting new methods, preserving consistency across many services, and avoiding typical blunders like accessing APIs or DB queries within loops. Sensitive data should not be uploaded, and changes to environment variables or database models should be executed consistently. Teams may work more effectively and develop higher-quality software by adhering to these standards.

## Type of change
Please choose appropriate options.

- [ ]  Bug fix (non-breaking change which fixes an issue)
- [ ]  New feature (non-breaking change which adds functionality)
- [ ]  Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ]  Enhancement (additive changes to improve performance)
- [ ]  This change requires a documentation update

## Checklist 

- [ ]  It's critical to avoid making needless file modifications in contributions, such as adding new lines, console logs, or additional spaces, to guarantee cleaner and more efficient code. Furthermore, eliminating unnecessary imports from a file might enhance code readability and efficiency.
- [ ]  Ensure that the pull request is assigned to the right base branch and that the development branch name contains the JIRA Task Id. Furthermore, each commit message should include the JIRA Task Id in the manner "ED-100: message".
- [ ]  Only update packages if it is mentioned and authorized in the design document, and make sure that you have the required permissions.
- [ ]  Avoid making API and database queries inside a loop as it can lead to performance issues and slow down the system.
- [ ]  When calling another function inside a given function, add comments explaining the purpose and meaning of the passed arguments and expected return values.
- [ ]  If adding a blank argument in a function, add a comment explaining the reason for the blank argument.
- [ ]  Before submitting a pull request, do a self-review of your code to ensure there are no conflicts with the base branch and all comments have been addressed.
- [ ]  Before merging a pull request, it's important to have other team members review it to catch any potential errors or issues
- [ ]  To maintain code integrity, it's important to remove all related changes when removing code during a code review.
- [ ]  If new constants, endpoints, or utility functions are introduced, it is important to check if they already exist in the service to avoid any duplication.
- [ ]  Whenever a new environment variable is added to a service, it's important to ensure that the necessary changes are made to related files such as ".env.sample" and "envVariables.js" to maintain consistency and avoid errors. Additionally, the new environment variable should be added to the devops repository to ensure that it is properly documented and accessible to the team.
- [ ]  When adding a new function to a service, it is important to document it with relevant information such as the name, parameters, and return value in a consistent format across all services. Additionally, if there are any changes to the API response, ensure that the documentation in the controllers is updated accordingly. 
- [ ]  Write a clear and concise commit message that describes the changes made.
- [ ]  Maintain consistent function signature and code across all services when adding a function to multiple services. Implement changes to database models in all services that use the same model. 
- [ ]  Use only let and const. Do not use var.
- [ ]  Make common functions for repetitive code blocks.
- [ ]  Avoid uploading sensitive information such as secret tokens or passwords in pull requests to ensure data security.
- [ ]  Maintain consistent indentation and spacing throughout the code.

### CODE REVIEW

File: generics/helpers/common_handler_v2.js [hunk 1]
```diff
@@ -0,0 +1,719 @@
...
+				i++
+			})
+		)
+
+		resolve([arrayOfChartData, projectData])
+	})
+}
+
+function dateDiffInDays(a, b) {
+	const utc1 = Date.UTC(a.getFullYear(), a.getMonth(), a.getDate())
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:433]: @MallanagoudaB this is a common function, check if similar function is present in utils , if not move it to utils. Do not use a,b values for params. Use meaning full variable name

File: generics/helpers/common_handler_v2.js [hunk 2]
```diff
@@ -0,0 +1,719 @@
...
+	const utc2 = Date.UTC(b.getFullYear(), b.getMonth(), b.getDate())
+	return Math.floor((utc2 - utc1) / (1000 * 60 * 60 * 24))
+}
+
+const createChart = async function (chartData, imgPath) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:438]: @MallanagoudaB function signature is missing!

File: generics/helpers/common_handler_v2.js [hunk 3]
```diff
@@ -0,0 +1,719 @@
...
+			)
+
+			return resolve(formData)
+		} catch (err) {
+			return reject(err)
+		}
+	})
+}
+
+const uploadPdfToCloud = async function (fileName, userId, folderPath) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:467]: @MallanagoudaB function signature is missing!

File: generics/helpers/common_handler_v2.js [hunk 4]
```diff
@@ -0,0 +1,719 @@
...
+		} catch (err) {
+			return resolve({
+				success: false,
+				message: err.message,
+			})
+		}
+	})
+}
+
+async function copyBootStrapFile(from, to) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:509]: @MallanagoudaB function signature is missing!

File: generics/helpers/common_handler_v2.js [hunk 5]
```diff
@@ -0,0 +1,719 @@
...
+		readCss.on('finish', function () {
+			return resolve(readCss)
+		})
+		readCss.on('error', function (err) {
+			return resolve(err)
+		})
+	})
+}
+
+async function getEntityReportChartObjects(data) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:522]: @MallanagoudaB function signature is missing!

File: generics/helpers/common_handler_v2.js [hunk 6]
```diff
@@ -0,0 +1,719 @@
...
+		await Promise.all(getChartObjects).then(function (response) {
+			chartData.push(response[0])
+			chartData.push(response[1])
+		})
+
+		return resolve(chartData)
+	})
+}
+
+async function getCategoryWiseChart(categories) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:537]: @MallanagoudaB function signature is missing!

File: generics/helpers/common_handler_v2.js [hunk 7]
```diff
@@ -0,0 +1,719 @@
...
+
+		let chartObject = {
+			order: 2,
+			options: chartOptions,
+		}
+		resolve(chartObject)
+	})
+}
+
+async function getTaskOverviewChart(tasks) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:618]: @MallanagoudaB function signature is missing!

File: generics/helpers/files_helper.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
+module.exports = {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/files_helper.js:1]: @MallanagoudaB we already have constants/common.js file. move these constants there, check for redundancy also

File: generics/helpers/utils.js [hunk 9]
```diff
@@ -3,363 +3,390 @@
...
-    parsedData._arrayFields.split(",").forEach(arrayTypeField => {
-      if (parsedData[arrayTypeField]) {
-        parsedData[arrayTypeField] = parsedData[arrayTypeField].split(",")
-      }
-    })
-  }
+	return parsedData
+}
+
+function getGotenbergConnection() {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:131]: @MallanagoudaB Why this is added in utils ? create a new file in generics/services/gotenberg.js and add this function there

File: generics/services/report.js [hunk 10]
```diff
@@ -7,151 +7,138 @@
 
 //dependencies
 
-const request = require('request');
-const reportsUrl = process.env.ML_REPORTS_SERVICE_URL;
+const request = require('request')
+const reportsUrl = process.env.ML_REPORTS_SERVICE_URL
 
 /**
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/report.js:13]: @MallanagoudaB do we need this file?

File: module/project/templateTasks/helper.js [hunk 11]
```diff
@@ -6,888 +6,799 @@
...
+ * ProjectTemplateTasksHelper
+ * @class
+ */
 
 // Dependencies
-const learningResourcesHelper = require(MODULES_BASE_PATH + "/learningResources/helper");
-const projectTemplateTaskQueries = require(DB_QUERY_BASE_PATH + "/projectTemplateTask");
-const projectTemplateQueries = require(DB_QUERY_BASE_PATH + "/projectTemplates");
-const solutionsQueries = require(DB_QUERY_BASE_PATH + "/solutions");
+const learningResourcesHelper = require(MODULES_BASE_PATH + '/learningResources/helper')
```
Comment (VISHNUDAS-tunerlabs) [inline module/project/templateTasks/helper.js:14]: @MallanagoudaB What is the change in this file or is it just formatting?

File: module/reports/helper.js [hunk 12]
```diff
@@ -7,736 +7,879 @@
 
 // Dependencies
 
-const userProjectsHelper = require(MODULES_BASE_PATH + "/userProjects/helper");
```
Comment (VISHNUDAS-tunerlabs) [inline module/reports/helper.js:10]: @MallanagoudaB What is the change in this file or is it just formatting?

File: module/solutions/helper.js [hunk 13]
```diff
@@ -5,1133 +5,1013 @@
...
 
-const { ObjectId } = require("mongodb");
+const { ObjectId } = require('mongodb')
 
 // Dependencies
 
-const solutionsQueries = require(DB_QUERY_BASE_PATH + "/solutions");
-const programsHelper = require(MODULES_BASE_PATH + "/programs/helper")
-const programQueries = require(DB_QUERY_BASE_PATH + "/programs")
+const solutionsQueries = require(DB_QUERY_BASE_PATH + '/solutions')
```
Comment (VISHNUDAS-tunerlabs) [inline module/solutions/helper.js:12]: @MallanagoudaB What is the change in this file or is it just formatting?

File: module/userProjects/helper.js [hunk 14]
```diff
@@ -3296,7 +3298,7 @@ function _entitiesInformation(entityIds, userToken) {
 
 			if (locationIds.length > 0) {
 				let queryData = {
-					'registryDetails.locationId': { $in: locationIds },
+					_id: { $in: locationIds },
```
Comment (VISHNUDAS-tunerlabs) [inline module/userProjects/helper.js:3301]: @MallanagoudaB What is this change?

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
Reviewed on May 10 4:50

File: controllers/v1/project/templateTasks.js [hunk 15]
```diff
@@ -6,138 +6,124 @@
...
-    * ProjectTemplateTasks
-    * @class
-*/
+/**
+ * ProjectTemplateTasks
+ * @class
+ */
 
 module.exports = class ProjectTemplateTasks extends Abstract {
-
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply controllers/v1/project/templateTasks.js:18]: i didnt made any changes there i think i pulled the code from prajwal so may be

File: generics/helpers/common_handler.js [hunk 16]
```diff
@@ -0,0 +1,559 @@
...
+const rimraf = require('rimraf')
+const ejs = require('ejs')
+const request = require('request')
+const rp = require('request-promise')
+const filesHelper = require(MODULES_BASE_PATH + '/cloud-services/files/helper')
+const axios = require('axios')
+
+exports.unnatiViewFullReportPdfGeneration = async function (responseData, userId) {
+	return new Promise(async function (resolve, reject) {
+		var currentTempFolder = 'tmp/' + uuidv4() + '--' + Math.floor(Math.random() * (10000 - 10 + 1) + 10)
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply generics/helpers/common_handler.js:17]: we are not calling any function

File: generics/helpers/common_handler.js [hunk 17]
```diff
@@ -0,0 +1,559 @@
...
+												if (
+													pdfDownloadableUrl.result[0] &&
+													Object.keys(pdfDownloadableUrl.result[0]).length > 0
+												) {
+													fs.readdir(imgPath, (err, files) => {
+														if (err) throw err
+
+														let i = 0
+														for (const file of files) {
+															fs.unlink(path.join(imgPath, file), (err) => {
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply generics/helpers/common_handler.js:104]: yes, its removed

File: module/project/templateTasks/helper.js (hunk 11 above)
Comment (MallanagoudaB) [reply module/project/templateTasks/helper.js:14]: i didnt added any thing there might be formatting

File: module/solutions/helper.js (hunk 13 above)
Comment (MallanagoudaB) [reply module/solutions/helper.js:12]: i think it was made by prajwal

File: module/userProjects/helper.js (hunk 14 above)
Comment (MallanagoudaB) [reply module/userProjects/helper.js:3301]: i got this in conflicts

File: generics/helpers/utils.js (hunk 9 above)
Comment (MallanagoudaB) [reply generics/helpers/utils.js:131]: it was there before and have the same formate in ML-REPORT

File: module/reports/helper.js (hunk 12 above)
Comment (MallanagoudaB) [reply module/reports/helper.js:10]: formatting

Comment (MallanagoudaB) [reply module/reports/helper.js:10]: i have made changes for entity API

File: module/userProjects/helper.js (hunk 14 above)
Comment (VISHNUDAS-tunerlabs) [reply module/userProjects/helper.js:3300]: @MallanagoudaB does this change required? check with @Prajwal17Tunerlabs

File: generics/services/report.js (hunk 10 above)
Comment (VISHNUDAS-tunerlabs) [reply generics/services/report.js:10]: @MallanagoudaB we don't need this file. we now don't have any dependencies with report service
Why you resolved this without actually resolving it?

File: generics/services/report.js [hunk 18]
```diff
@@ -7,151 +7,135 @@
...
-  * @name viewFullReport
-  * @param {String} token - user token for verification 
-  * @param {String} input - input request body
-  * @returns {JSON} - consist of pdf url
-*/
-
-const viewFullReport = function (token,input) {
-    return new Promise(async (resolve, reject) => {
-        try {
-            const url =
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/report.js:25]: @MallanagoudaB remove this file and its imports from code base

File: generics/helpers/common_handler.js [hunk 19]
```diff
@@ -0,0 +1,437 @@
...
+	})
+}
+
+/**
+ * Calculate the difference in days between two dates.
+ * @param {Date} a - The first date.
+ * @param {Date} b - The second date.
+ * @returns {number} The difference in days between the two dates (b - a).
+ */
+function dateDiffInDays(startDate, endDate) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler.js:317]: @MallanagoudaB this is a common function move it to utils and call from there

File: generics/helpers/common_handler_v2.js [hunk 20]
```diff
@@ -0,0 +1,848 @@
...
+	})
+}
+
+/**
+ * Calculate the difference in days between two dates.
+ * @param {Date} a - The first date.
+ * @param {Date} b - The second date.
+ * @returns {number} The difference in days between the two dates (b - a).
+ */
+function dateDiffInDays(startDate, endDate) {
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/common_handler_v2.js:498]: @MallanagoudaB Why duplicate this function? once you add the function to utils you can call it directly where ever it is used.

File: generics/helpers/utils.js [hunk 21]
```diff
@@ -539,36 +558,35 @@ function checkIfValidUUID(value) {
...
-};
+	camelCaseToTitleCase: camelCaseToTitleCase,
+	lowerCase: lowerCase,
+	checkIfStringIsUrl: checkIfStringIsUrl,
+	hyphenCaseToCamelCase: hyphenCaseToCamelCase,
+	valueParser: valueParser,
+	convertStringToBoolean: convertStringToBoolean,
+	getAllBooleanDataFromModels: getAllBooleanDataFromModels,
+	epochTime: epochTime,
+	getGotenbergConnection: getGotenbergConnection,
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:573]: @MallanagoudaB Please move this function to a new file and call it there.path :  generics/services/gotenberg.js

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on 15 may 7:42

File: app.js [hunk 22]
```diff
@@ -5,67 +5,72 @@
  * Description : Root file.
  */
 
-require("dotenv").config();
+require('dotenv').config()
```
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply app.js:8]: okay vishnu

File: module/userProjects/helper.js (hunk 14 above)
Comment (MallanagoudaB) [reply module/userProjects/helper.js:3300]: @Prajwal17Tunerlabs added this

File: controllers/v1/reports.js [hunk 23]
```diff
@@ -58,47 +57,43 @@ module.exports = class Reports {
...
+					status: error.status || HTTP_STATUS_CODE.internal_server_error.status,
+					message: error.message || HTTP_STATUS_CODE.internal_server_error.message,
+					errorObject: error,
+				})
+			}
+		})
+	}
+
+	/**
     * @api {post} /improvement-project/api/v1/reports/getProgramsByEntity/:_id
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reports.js:97]: @MallanagoudaB when you make change, look for similar changes please. correct the url

File: controllers/v1/reports.js [hunk 24]
```diff
@@ -134,48 +129,45 @@ module.exports = class Reports {
...
+				return reject({
+					status: error.status || HTTP_STATUS_CODE.internal_server_error.status,
+					message: error.message || HTTP_STATUS_CODE.internal_server_error.message,
+					errorObject: error,
+				})
+			}
+		})
+	}
+	/**
     * @api {get} /improvement-project/api/v1/reports/detailView/:_id?requestPdf=:requestPdf&programId=:programId&reportType=:reportType
```
Comment (VISHNUDAS-tunerlabs) [inline controllers/v1/reports.js:166]: @MallanagoudaB correct here also

File: generics/constants/common.js [hunk 25]
```diff
@@ -76,4 +76,9 @@ module.exports = {
 	READ_PERMISSION: 'r',
 	PUBLIC: 'public',
 	PRIVATE: 'private',
+	AZURE: 'azure',
+	SUCCESS: 'SUCCESS',
```
Comment (VISHNUDAS-tunerlabs) [inline generics/constants/common.js:80]: @MallanagoudaB already one success present as constant, do not duplicate

File: module/userProjects/helper.js [hunk 26]
```diff
@@ -1944,15 +1944,15 @@ module.exports = class UserProjectsHelper {
 				if (UTILS.revertStatusorNot(appVersion)) {
 					projectDocument.status = UTILS.revertProjectStatus(projectDocument.status)
 				}
-				let response = await reportService.projectAndTaskReport(userToken, projectDocument, projectPdf)
 
-				if (response && response.success == true) {
+				let response = await common_handler.unnatiViewFullReportPdfGeneration(projectDocument, userToken)
```
Comment (VISHNUDAS-tunerlabs) [inline module/userProjects/helper.js:1948]: @MallanagoudaB Why are you passing user token instead user id? I have added this comment in prev review also

File: module/userProjects/helper.js (hunk 14 above)
Comment (VISHNUDAS-tunerlabs) [reply module/userProjects/helper.js:3299]: @MallanagoudaB my question is which one is correct?

File: generics/helpers/common_handler_v2.js [hunk 27]
```diff
@@ -0,0 +1,848 @@
...
+	})
+}
+
+/**
+ * Generates a full PDF report containing Gantt chart visualizations and uploads it to cloud storage.
+ * @param {Object} responseData - Data object for generating the report.
+ * @param {string} userId - User ID for identifying the uploader.
+ * @returns {Promise<Object>} A promise that resolves to an object representing the result of PDF generation and upload.
+ */
+exports.unnatiViewFullReportPdfGeneration = async function (responseData, userId) {
```
[the remark this answers is not in the document]
Comment (VISHNUDAS-tunerlabs) [reply generics/helpers/common_handler_v2.js:203]: @MallanagoudaB I am expecting a reply here

File: generics/helpers/utils.js [hunk 28]
```diff
@@ -504,33 +510,43 @@ function generateUniqueId() {
...
+		HOSTNAME: '',
+		'application.home': '',
+	}
+	return telemetryEvent
 }
 
-
+/**
+ * Calculate the difference in days between two dates.
+ * @param {Date} a - The first date.
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:527]: @MallanagoudaB variable name is incorrect

File: generics/helpers/utils.js [hunk 29]
```diff
@@ -504,33 +510,43 @@ function generateUniqueId() {
...
+ * @param {Date} b - The second date.
```
Comment (VISHNUDAS-tunerlabs) [inline generics/helpers/utils.js:528]: @MallanagoudaB correct the variable name

File: generics/services/gotenberg.js [hunk 30]
```diff
@@ -0,0 +1,23 @@
+/**
+ * Generates options for making a request to Gotenberg service to convert HTML content.
+ * @returns {Object} The options object for the HTTP request.
+ */
+function getGotenbergConnection() {
+	let options = {
+		method: 'POST',
+		uri: process.env.GOTENBERG_URL + '/forms/chromium/convert/html',
```
Comment (VISHNUDAS-tunerlabs) [inline generics/services/gotenberg.js:8]: @MallanagoudaB add this new env to env samples and envVariables.js file

File: module/reports/helper.js [hunk 31]
```diff
@@ -7,736 +7,884 @@
...
+		})
+	}
+
+	/**
+	 * Generates a PDF report based on entity report data.
+	 * @param {Object} entityReportData - Data for generating the report.
+	 * @param {string} userId - User ID associated with the report generation.
+	 * @returns {Promise<Object>} A promise that resolves to an object containing PDF generation status and URL.
+	 */
+	unnatiEntityReportPdfGeneration = async function (entityReportData) {
```
Comment (VISHNUDAS-tunerlabs) [inline module/reports/helper.js:353]: @MallanagoudaB is this function similar to the one we have in common_handler. Where this function used?

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed on may 16 - 3 PM

File: module/userProjects/helper.js (hunk 14 above)
Comment (MallanagoudaB) [reply module/userProjects/helper.js:3299]: 'registryDetails.locationId': { $in: locationIds }, this should be present and code is not braking anywere bcause of this

File: generics/helpers/common_handler_v2.js (hunk 27 above)
[the remark this answers is not in the document]
Comment (MallanagoudaB) [reply generics/helpers/common_handler_v2.js:203]: Im passing userID only @VISHNUDAS-tunerlabs  in helper

File: module/reports/helper.js (hunk 31 above)
Comment (MallanagoudaB) [reply module/reports/helper.js:353]: its the same function i removed it and no code break after this

File: module/userProjects/helper.js (hunk 14 above)
Comment (VISHNUDAS-tunerlabs) [reply module/userProjects/helper.js:3300]: @MallanagoudaB that is not the present code right?

Comment (MallanagoudaB) [reply module/userProjects/helper.js:3300]: no its not there

Comment (VISHNUDAS-tunerlabs) [reply module/userProjects/helper.js:3300]: @MallanagoudaB then we should keep the old logic

File: .env.sample [hunk 32]
```diff
@@ -60,4 +60,6 @@ APP_PORTAL_BASE_URL = "https://dev.elevate.org"
 
 PRESIGNED_URL_EXPIRY_IN_SECONDS = 300                                           // presigned url expiry in seconds after which the signed url becomes invalid
 
-DOWNLOADABLE_URL_EXPIRY_IN_SECONDS = 300                                        // downloadable url expiry in seconds after which the downloadable url becomes invalid 
\ No newline at end of file
+DOWNLOADABLE_URL_EXPIRY_IN_SECONDS = 300                                        // downloadable url expiry in seconds after which the downloadable url becomes invalid 
+
+GOTENBERG_URL=http://localhost:3000
```
Comment (VISHNUDAS-tunerlabs) [inline .env.sample:65]: @MallanagoudaB give explanation comment

**Review by VISHNUDAS-tunerlabs — CHANGES_REQUESTED** [review]
reviewed IV

File: module/userProjects/helper.js (hunk 14 above)
Comment (MallanagoudaB) [reply module/userProjects/helper.js:3300]: that changes was made by @Prajwal17Tunerlabs  we need to confirm with him

Comment (VISHNUDAS-tunerlabs) [reply module/userProjects/helper.js:3301]: @MallanagoudaB  in the second comment its self I told you to check with him and confirm. Why you are not doing it still? connect with him and give me the correct query. Please don't make me repeat this again

**Pull request merged**